opencog/query/Distributed/DistSCM.cc

```diff
@@ -0,0 +1,230 @@
+/*
+ * DistSCM.cc
+ *
+ *
+ * Copyright (C) 2015 OpenCog Foundation
+ *
+ * Author: Mandeep Singh Bhatia , September 2015
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License v3 as
+ * published by the Free Software Foundation and including the exceptions
+ * at http://opencog.org/wiki/Licenses
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program; if not, write to:
+ * Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#define MAX_SCM_CODE_STRING 5000
+
+#include "DistSCM.h"
+#include <opencog/guile/SchemeEval.h>//which lib to link?
+#include <opencog/guile/SchemeSmob.h>
+
+using namespace opencog;
+
+bool DistSCM::master_mode=true;
+
+DistSCM::DistSCM(void):true_string("true"),false_string("false")
+{
+  scm_with_guile(init_in_guile, this);
+}
+
+void* DistSCM::init_in_guile(void* self)
+{
+    scm_c_define_module("opencog dist-gearman", init_in_module, self);
+    scm_c_use_module("opencog dist-gearman");
+    return NULL;
+}
+
+void DistSCM::init_in_module(void* data)
+{
+    DistSCM* self = (DistSCM*) data;
+    self->init();
+}
+
+void DistSCM::init(void)
+{
+  //blocking slave mode
+	define_scheme_primitive("set-slave-mode", &DistSCM::slave_mode,
+                            this, "dist-gearman");
+  //exits the slave mode after last request is served
+	define_scheme_primitive("set-master-mode", &DistSCM::set_master_mode,
+                            this, "dist-gearman");
+  //sends scheme string to slave and blocks till result arrives, returns uuid of resulting handle, invalid_uuid on failure
+  define_scheme_primitive("dist-run-scm", &DistSCM::dist_scm,
+                            this, "dist-gearman");
+}
+
+void DistSCM::set_master_mode(void)
+{
+	master_mode=true;
+}
+
+gearman_return_t DistSCM::worker_function(gearman_job_st *job, void *context)
+{
+  const char *workload= (const char *)gearman_job_workload(job);
+  const size_t workload_size= gearman_job_workload_size(job);
+
+  //std::cout << "Recieved " << workload_size << " bytes" << std::endl;
+  if (workload_size>MAX_SCM_CODE_STRING+1) return GEARMAN_ERROR;
+  char scm_str[MAX_SCM_CODE_STRING+1];
+  strcpy(scm_str,workload);
+
+  SchemeEval evl((AtomSpace*)context);
+  evl.init_scheme();
+  //std::cout<<scm_str<<"\n";
+  Handle result=evl.eval_h(scm_str);//can we check if handle is valid? how to get current atomspace
+  UUID vl=result.value();
+  if (Handle::UNDEFINED==result)std::cerr<<"invalid handle \n";
+
+  evl.eval("(sql-store)");
```

Comment:
actually, I'm thinking that you should not do sql-anything at all, in this code. Let the scheme code that is running in the workers decide if it needs to call sql.  If the worker found zero atoms,, its pointless to try to "store" them. The worker is going to have a much better idea of what the correct thing to do is, than the dist framework will.